fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,748 @@
+# Copyright (c) Fairlearn contributors.
+# Licensed under the MIT License.
+
+from math import ceil
+from time import time
+from ._constants import (
+    _IMPORT_ERROR_MESSAGE,
+    _KWARG_ERROR_MESSAGE,
+    _PREDICTION_FUNCTION_AMBIGUOUS,
+    _PROGRESS_UPDATE,
+)
+from ._backend_engine import BackendEngine
+from ._pytorch_engine import PytorchEngine
+from ._tensorflow_engine import TensorflowEngine
+from ._preprocessor import (
+    FloatTransformer,
+    _get_type,
+)
+from sklearn.base import (
+    ClassifierMixin,
+    RegressorMixin,
+    BaseEstimator,
+    TransformerMixin,
+)
+from sklearn.utils.validation import (
+    check_is_fitted,
+    check_random_state,
+    check_array,
+)
+from sklearn.exceptions import NotFittedError
+from numpy import zeros, argmax, arange
+
+
+class AdversarialFairness(BaseEstimator):
```

Comment:
This name seems (to me) like it's missing a part, like `estimator` or `mitigator`.

Reply:
Yes! Excellent point. AdversarialFairnessEstimator or AdversarialMitigator or so, what do you like?

Reply:
Given `AdversarialFairnessClassifier` and `AdversarialFairnessRegressor` I think `AdversarialFairnessEstimator` makes the most sense, but I'm happy either way.

Reply:
Actually, another option would be to make `AdversarialFairnessEstimator` a private class and only export `AdversarialFairnessClassifier` and `AdversarialFairnessRegressor`. I can see arguments either way. Do folks have any thoughts on this?

Reply:
Rather not, because we are purposefully exposing the losses and other API to make it applicable to more difficult use cases that do not fall under classification or regression, such as multi-output/multitask classification

Reply:
I thought about all of this some more, and I'm still finding parts of the constructor API unnecessarily complicated--and there are various dependencies spread across multiple parameters. In particular, I'd like to suggest an alternative to what's currently handled by: `predictor_loss, adversary_loss, predictor_function, y_transform, sf_transform`

**AdversarialMitigationClassifier**
* can handle binary classification and multi-class classification
* for encoding of `y`, we follow sklearn`s [target types](https://scikit-learn.org/stable/glossary.html#glossary-target-types), which are described [here](https://scikit-learn.org/stable/modules/multiclass.html)
* _Binary classification_
  * `y` has shape `(n_samples,)` and it contains two values typically {0,1}; if other values are provided they are transformed into {0,1} during fit and predict
  * the neural net is outputting a single scalar, corresponding to the logit of P(Y=1|X), and is trained by minimizing logistic loss
 * _Multiclass classification_
   * `y` has either shape `(n_samples,n_classes)` with 0/1 values implementing one-hot-encoding, or it has shape `(n_samples,)` and contains 3 or more distinct values; in the latter case, it is transformed into one-hot-encoding during fit and predict
   * the neural net is outputting a vector of `n_classes` scalars, corresponding to the linear scores of multinomial logistic model of P(Y|X), and is trained by minimizing log loss (this is called CrossEntropyLoss in PyTorch)
* _Adversarial model_
   * sensible defaults around fitting sensitive features are a bit more tricky, but I think that we should begin by covering the common case of binary or categorical sensitive features:
     * `sensitive_features` is shape `(n_samples,)` or `(n_samples,n_sensitive_features)` where every column has two distinct values -> columns are transformed into {0,1}, and the training loss is the sum of logistic losses (i.e., treated as the sum of binary problems)
     * `sensitive_features` is shape `(n_samples,)` or `(n_samples,n_sensitive_features)` where at least one column has three or more values -> columns are transformed using one-hot-encoding, and the training loss is the sum of log losses (i.e., treated as the sum of multiclass problems)
* **This means that we could get rid of the parameters `predictor_loss, adversary_loss, predictor_function, y_transform, sf_transform`.**
  * If we want to expose more generality (say continuous sensitive features), we could do that later.

**AdversarialMitigationRegressor**
* `y` has shape `(n_samples,)` and contains floats, the loss is square loss
* sensitive features and adversarial model are treated the same way as in AdversarialMitigationClassifier.
* again, this means we could get rid of `predictor_loss, adversary_loss, predictor_function, y_transform, sf_transform`, and add more generality later.

**AdversarialMitigation**
* I suggest to make this private; in that case I don't care too much about the API.